libraries/ms-common/src/x86/can_hw.c

```diff
@@ -1,27 +1,202 @@
+  struct can_filter filters[CAN_HW_MAX_FILTERS];
+  size_t num_filters;
+  CANHwEventHandler handlers[NUM_CAN_HW_EVENTS];
+  uint32_t delay_us;
+} CANHwSocketData;
+
+static pthread_t s_rx_pthread_id;
+static pthread_t s_tx_pthread_id;
+static pthread_mutex_t s_tx_mutex = PTHREAD_MUTEX_INITIALIZER;
+static pthread_cond_t s_tx_cond = PTHREAD_COND_INITIALIZER;
+
+static CANHwSocketData s_socket_data = {.can_fd = -1 };
+
+static uint32_t prv_get_delay(CANHwBitrate bitrate) {
+  uint32_t delay_us[NUM_CAN_HW_BITRATES] = {
+    1000,  // 125 kbps
+    500,   // 250 kbps
+    250,   // 500 kbps
+    125,   // 1 mbps
+  };
+
+  return delay_us[bitrate];
+}
+
+static void *prv_rx_thread(void *arg) {
+  while (true) {
+    int bytes = read(s_socket_data.can_fd, &s_socket_data.rx_frame, sizeof(s_socket_data.rx_frame));
+
+    if (s_socket_data.handlers[CAN_HW_EVENT_MSG_RX].callback != NULL) {
+      s_socket_data.handlers[CAN_HW_EVENT_MSG_RX].callback(
+          s_socket_data.handlers[CAN_HW_EVENT_TX_READY].context);
+    }
+
+    // Limit how often we can receive messages to simulate bus speed
+    usleep(s_socket_data.delay_us);
+  }
+
+  return NULL;
+}
+
+static void *prv_tx_thread(void *arg) {
+  struct can_frame frame = { 0 };
+
+  while (true) {
+    pthread_mutex_lock(&s_tx_mutex);
+    while (fifo_size(&s_socket_data.tx_fifo) == 0) {
+      pthread_cond_wait(&s_tx_cond, &s_tx_mutex);
+    }
+
+    fifo_pop(&s_socket_data.tx_fifo, &frame);
+    pthread_mutex_unlock(&s_tx_mutex);
+
+    int bytes = write(s_socket_data.can_fd, &frame, sizeof(frame));
+
+    // Delay to simulate bus speed
+    usleep(s_socket_data.delay_us);
+
+    if (s_socket_data.handlers[CAN_HW_EVENT_TX_READY].callback != NULL) {
+      s_socket_data.handlers[CAN_HW_EVENT_TX_READY].callback(
+          s_socket_data.handlers[CAN_HW_EVENT_TX_READY].context);
+    }
+  }
+
+  return NULL;
+}
 
 StatusCode can_hw_init(const CANHwSettings *settings) {
-  return status_code(STATUS_CODE_UNIMPLEMENTED);
+  if (s_socket_data.can_fd != -1) {
+    // Reinit everything
+    // This should eventually be done properly with signals
+    pthread_cancel(s_rx_pthread_id);
+    pthread_cancel(s_tx_pthread_id);
+    close(s_socket_data.can_fd);
+  }
+
+  pthread_mutex_init(&s_tx_mutex, NULL);
+  pthread_cond_init(&s_tx_cond, NULL);
+
+  memset(&s_socket_data, 0, sizeof(s_socket_data));
+  s_socket_data.delay_us = prv_get_delay(settings->bitrate);
+  fifo_init(&s_socket_data.tx_fifo, s_socket_data.tx_frames);
+
+  s_socket_data.can_fd = socket(PF_CAN, SOCK_RAW, CAN_RAW);
+  if (s_socket_data.can_fd == -1) {
+    LOG_CRITICAL("CAN HW: Failed to open socketcan socket\n");
+    return status_code(STATUS_CODE_INTERNAL_ERROR);
+  }
+
+  LOG_DEBUG("CAN HW initialized on %s\n", CAN_HW_DEV_INTERFACE);
+  // Loopback - expects to receive its own messages
+  int loopback = settings->loopback;
+  setsockopt(s_socket_data.can_fd, SOL_CAN_RAW, CAN_RAW_RECV_OWN_MSGS, &loopback, sizeof(loopback));
```

Comment:
Should probably check this return value and the add something along the lines of:
```c
if (err) {
  LOG_DEBUG("setscokopt error: %s\n", strerror(errno));
  return status_code(STATUS_CODE_INTERNAL_ERROR);
}
```

Which also reminds me that maybe in `setup_test` calls we should always add a status callback to handle messages/codes and make debugging more transparent in tests. Thoughts? If you agree I'll make a ticket for it.

Reply:
Hmm, I worry that adding a status callback in every test would add a lot of noise when we run our tests. There's a combination of tests that purposely cause errors and modules that will continuously cause errors (e.g. event loops)

Reply:
Good point. That makes sense. Regardless I might add a generic status handler fn pointer that just `LOG_DEBUG`s the errors so it is faster to configure one when debugging? But I won't actually call it anywhere It would just be another type of debug util.

Reply:
Makes sense to me.